### PR TITLE
libnet: Don't forward to upstream resolvers on internal nw

### PR DESCRIPTION
**Pull request by akerouanton**
- Stacked on top of #46603
- Related to CI failures seen on #46531

**- What I did**

First commit adds a way to enable/disable upstream forwarding to the embedded resolver. When an endpoint joins/leaves a sandbox, this forwarding policy is modified based on whether there's an endpoint providing external connectivity.

Second commit ensures the resolver return a REFUSED response with the RA bit (ie. Recursion Available) unset when upstream forwarding is disabled and no matching container/alias can be found.

**- Description for the changelog**

* The embedded DNS resolver now returns a REFUSED response when upstream forwarding is disabled (ie. on internal networks, or on Windows) and no container/alias match the queried name.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/resolver_test.go [hunk 1]
```diff
@@ -260,10 +260,6 @@ func TestReplySERVFAIL(t *testing.T) {
 			name: "InternalError",
 			q:    new(dns.Msg).SetQuestion("_sip._tcp.example.com.", dns.TypeSRV),
 		},
-		{
-			name: "ProxyDNS=false",
-			q:    new(dns.Msg).SetQuestion("example.com.", dns.TypeA),
-		},
 		{
 			name:     "ProxyDNS=true", // No extDNS servers configured -> no answer from any upstream
```
Comment (akerouanton) [inline libnetwork/resolver_test.go:264]: I'm wondering if we should return a proper NXDOMAIN in that case too :thinking:

File: libnetwork/resolver.go [hunk 2]
```diff
@@ -75,7 +76,7 @@ type Resolver struct {
 	tcpListen     *net.TCPListener
 	err           error
 	listenAddress string
-	proxyDNS      bool
+	proxyDNS      *atomic.Bool
```
Comment (corhere) [inline libnetwork/resolver.go:79]: does `Resolver` need to be copyable?
```suggestion
	proxyDNS      atomic.Bool
```

**Comment by neersighted** [comment]
Closing as this was carried in https://github.com/moby/moby/pull/47538.

**Comment by ArchanaWind** [comment]
HI,
CVE-2024-29018 is created for this issue, is moby_20.10.x is vulnerable to this bug


Thanks in advance

**Pull request closed without merging**